---
title: 如何优雅地处理前端异常
comments: true
date: 2019-08-11 10:00:00
categories: JS
tags: ['JS']
---

> 前端是最贴近用户的一层，随着产品的日益完善，越来越注重用户体验，前端异常就显得非常必要。


### 为什么要处理异常？
异常是不可控的，会影响最终的呈现结果，但是我们有充分的理由去做这样的事情。
1.增强用户体验；
2.远程定位问题；
3.未雨绸缪，及早发现问题；
4.无法复现问题，尤其是移动端，机型，系统都是问题；
5.完善的前端方案，前端监控系统；
6.提升找错效率，提升前端代码安全性。

对于 JS 而言，我们面对的仅仅只是异常，异常的出现不会直接导致 JS 引擎崩溃，最多只会使当前执行的任务终止。


### 前端异常捕获与错误上报
前端监控包括行为监控、异常监控、性能监控等。一个监控系统，可以分为四个阶段：日志采集、日志存储、统计与分析、报告和警告。
异常采集内容 4W: WHO did WHAT and get WHICH exception in WHICH environment?
浏览器信息 navigator.userAgent获取，device设备、browser浏览器、version浏览器版本、engine内核、os操作系统、Version操作系统版本、language语言


### 前端需要处理哪些异常？
* JS 错误
SyntaxError    语法错误 - 解析代码时发生的语法错误。
ReferenceError 引用错误 - 引用一个不存在的变量时发生的错误。
RangeError     范围错误 - 当一个值超出有效范围时发生的错误。
TypeError      类型错误 - 是变量或参数不是预期类型时发生的错误。

* Ajax 请求异常
* 静态资源加载异常
* Promise 异常
* Iframe 异常
* 跨域 script error
* 崩溃和卡顿


#### 异常捕获方法

* try catch 
在可预见情况下监控特定的错误。只能捕获到同步的运行时错误，对语法和异步错误却无能为力

* window.onerror
捕获预料之外的错误，全局监听异常，但不能捕获语法错误和promise的异常错误信息。最好写在所有 JS 脚本的前面，否则有可能捕获不到错误。

script error跨域时 script 加上属性crossorigin，并且服务器添加 Access-Control-Allow-Origin。

```js
/**
* @param {String}  message    错误信息
* @param {String}  source    出错文件
* @param {Number}  lineno    行号
* @param {Number}  colno    列号
* @param {Object}  error  Error对象（对象）
*/

window.onerror = function(message, source, lineno, colno, error) {
   console.log('捕获到异常：',{message, source, lineno, colno, error})
}
```

* window.addEventListener 的 error 事件
当资源（如图片或脚本）加载失败，加载资源的元素会触发一个 Event 接口的 error 事件，并执行该元素上的onerror() 处理函数。这些 error 事件不会向上冒泡到 window。

```js
window.addEventListener('error', (error) => {
    console.log('捕获到异常：', error);
}, true)
```

由于网络请求异常不会事件冒泡，因此必须在捕获阶段将其捕捉到才行，但是这种方式虽然可以捕捉到网络请求的异常，但是无法判断 HTTP 的状态是 404 还是其他比如 500 等等，所以还需要配合服务端日志进行排查分析。


* Promise Catch
在 promise 中使用 catch 可以非常方便的捕获到异步 error


* window.addEventListener 的 unhandledrejection、handledrejection 事件
捕获到未处理的Promise错误。
没有写 catch 的 Promise 中抛出的错误无法被 onerror 或 try-catch 捕获到，为了防止有漏掉的 Promise 异常，在全局增加一个对 unhandledrejection 的监听，用来全局监听 Uncaught Promise Error。

```js
window.addEventListener("unhandledrejection", function(e){
  e.preventDefault()  // 有这行时，会去掉控制台的异常显示
  console.log('捕获到异常：', e);
  return true;
});
Promise.reject('promise error');
```

* aixos 进行拦截
aixos 中使用 interceptor 进行拦截
```js
// http request 拦截器
axios.interceptors.request.use(config => {
    //loading
    if (store.state.user.token) {
        config.headers['Authorization'] = store.state.user.token //每个请求都带上token
        // config.headers['Cookie'] = myCookie //自定义Cookie
    }
      return config
    }, err => {
      Message.error({message: '请求超时!'});
      return Promise.reject(err)
    }
)

//http response 拦截器
axios.interceptors.response.use(
    response => {
      if (response.data.code === 5555) {
            Message.error({
                message: '出现非法信息: ' + response.data.illegal
            })
            return Promise.reject(response.data.description)
        } else {
            return response
        }
    }, 
    err => {
        if (err.response) {
            switch (err.response.status) {
              // 401 清除token信息并跳转到登录页面
              case 401:
                  err.message = 'Unauthorized 未授权，请重新登录'
                  store.commit('LogOut')
                  router.replace({ path: '/login'})
                  break
              case 403:
                  err.message = 'Forbidden 服务器拒绝访问'
                  break
              default:
                  err.message = `连接错误${err.response.status}`
            }
        }
        console.log(err)
        return Promise.reject(err)
    })

```

* VUE errorHandler
vue 使用 Vue.config.errorHandler 捕获全局异常
```js
Vue.config.errorHandler = (err, vm, info) => {
  console.error('通过vue errorHandler捕获的错误');
  console.error(err);
  console.error(vm);
  console.error(info);
}
```

* React 异常捕获
React 16 提供了一个内置函数 componentDidCatch，使用它可以非常简单的获取到 react 下的错误信息

```js
componentDidCatch(error, info) {
    console.log(error, info);
}
```


此外，UI 的某部分引起的 JS 错误不应该破坏整个程序，React16 提出了错误边界（error boundary)的新观念。但不会捕捉 事件处理器、异步代码、服务端的渲染代码、在 error boundaries 区域内的错误。

```js
class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }
 
  componentDidCatch(error, info) {
    this.setState({ hasError: true });  // Display fallback UI

    logErrorToMyService(error, info);   // 将异常信息上报给服务器
  }
 
  render() {
    if (this.state.hasError) {
      return <h1>Something went wrong.</h1>;
    }
    return this.props.children;
  }
}

// 像使用普通组件那样使用
<ErrorBoundary>
  <MyWidget />
</ErrorBoundary>
```
componentDidCatch() 方法像 JS 的 catch{} 模块一样工作，但是对于组件，只有 class 类型的组件(class component )可以成为一个 error boundaries。实际上，大多数情况下我们可以在整个程序中定义一个 error boundary 组件，之后就可以一直使用它了！


* iframe 异常
对于 iframe 的异常捕获，我们还得借力 window.onerror
```html
<iframe src="./iframe.html" frameborder="0"></iframe>
<script>
  window.frames[0].onerror = function (message, source, lineno, colno, error) {
    console.log('捕获到 iframe 异常：',{message, source, lineno, colno, error});
    return true;
  };
</script>
```

* script error
script error跨域时 script 加上属性crossorigin='anonymous'，并且服务器添加 Access-Control-Allow-Origin。


* sourceMap  
将代码压缩发布的代码，无法找到原始报错位置。在webpack中的devtool开启sourceMap功能追踪错误，将异常信息提交到 node 中间层处理分析，解析map文件后将数据传给后台服务器。
```js
// 中间层代码
const express = require('express');
const fs = require('fs');
const router = express.Router();
const fetch = require('node-fetch');
const sourceMap = require('source-map');
const path = require('path');
const resolve = file => path.resolve(__dirname, file);

// 定义post接口
router.post('/errorMsg/', function(req, res) {
    let error = req.body; // 获取前端传过来的报错对象
    let url = error.scriptURI; // 压缩文件路径

    if (url) {
        let fileUrl = url.slice(url.indexOf('client/')) + '.map'; // map文件路径

        // 解析sourceMap
        let smc = new sourceMap.SourceMapConsumer(fs.readFileSync(resolve('../' + fileUrl), 'utf8')); // 返回一个promise对象
        
        smc.then(function(result) {
            // 解析原始报错数据
            let ret = result.originalPositionFor({
                line: error.lineNo, // 压缩后的行号
                column: error.columnNo // 压缩后的列号
            });
            
            let url = ''; // 上报地址
        
            // 将异常上报至后台
            fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    errorMessage: error.errorMessage, // 报错信息
                    source: ret.source, // 报错文件路径
                    line: ret.line, // 报错文件行号
                    column: ret.column, // 报错文件列号
                    stack: error.stack // 报错堆栈
                })
            }).then(function(response) {
                return response.json();
            }).then(function(json) {
                res.json(json);         
            });
        })
    }
});

module.exports = router;
```

* 崩溃和卡顿
利用 window 对象的 load 和 beforeunload 事件实现了网页崩溃的监控。
```js
window.addEventListener('load', function () {
    sessionStorage.setItem('good_exit', 'pending');
    setInterval(function () {
        sessionStorage.setItem('time_before_crash', new Date().toString());
    }, 1000);
  });

  window.addEventListener('beforeunload', function () {
    sessionStorage.setItem('good_exit', 'true');
  });

  if(sessionStorage.getItem('good_exit') &&
    sessionStorage.getItem('good_exit') !== 'true') {
    // insert crash logging code here ...
    alert('Hey, welcome back from your crash, looks like you crashed on: ' + sessionStorage.getItem('time_before_crash'));
  }
```

* Service Worker 
可以使用 Service Worker 来实现网页崩溃的监控。

Service Worker 有自己独立的工作线程，与网页区分开，网页崩溃了，Service Worker 一般情况下不会崩溃；
Service Worker 生命周期一般要比网页还要长，可以用来监控网页的状态；
网页可以通过 navigator.serviceWorker.controller.postMessage API 向掌管自己的 SW 发送消息。

基于以上几点，可以实现一种基于心跳检测的监控方案。


### 错误上报
* ajax传递给接口
因为 Ajax 请求本身也有可能会发生异常，而且有可能会引发跨域问题，一般情况下更推荐使用动态创建 img 标签的形式进行上报。


* 动态创建 img 标签
new Image().src=reportUrl?logs=error 的方式进行上报
```js
function report(error) {
  let reportUrl = 'http://supremeyh/web/report';
  new Image().src = `${reportUrl}?logs=${error}`;
}

Reporter.send = function(data) {
  // 设置采集率，减缓服务器的压力
  if(Math.random() < 0.3) {
    send(data)      // 上报错误信息
  }
}
```

### 节流
当数据量变大的时候，当一个异常一段时间内不断触发，那么上报的日志会造成数据冗余，也造成流量浪费。所以，对于错误信息的上报，从上报内容和上报频率上，应该加以限制。
1、限制上报：不是所有的错误都上报。设置采集率，添加黑白名单，或者添加过滤条件，满足条件的的才上报。
2、合并上报：一段时间内统一的错误汇总上报，减少请求。
3、服务端限制：客户端的东西总是可以绕开，服务端需要做限制，防止被攻击，或者负载过高。




